librz/bin/format/mach0/mach0_relocs.c

```diff
@@ -0,0 +1,645 @@
+
+#include <rz_util.h>
+#include "mach0.h"
+#include <ht_uu.h>
+
+#include "mach0_utils.inc"
+
+static int reloc_comparator(struct reloc_t *a, struct reloc_t *b) {
+	return a->addr - b->addr;
+}
+
+static void parse_relocation_info(struct MACH0_(obj_t) * bin, RzSkipList *relocs, ut32 offset, ut32 num) {
+	if (!num || !offset || (st32)num < 0) {
+		return;
+	}
+
+	ut64 total_size = num * sizeof(struct relocation_info);
+	struct relocation_info *info = calloc(num, sizeof(struct relocation_info));
+	if (!info) {
+		return;
+	}
+
+	if (rz_buf_read_at(bin->b, offset, (ut8 *)info, total_size) < total_size) {
+		free(info);
+		return;
+	}
+
+	size_t i;
+	for (i = 0; i < num; i++) {
+		struct relocation_info a_info = info[i];
+		ut32 sym_num = a_info.r_symbolnum;
+		if (sym_num > bin->nsymtab) {
+			continue;
+		}
+
+		ut32 stridx = bin->symtab[sym_num].n_strx;
+		char *sym_name = MACH0_(get_name)(bin, stridx, false);
+		if (!sym_name) {
+			continue;
+		}
+
+		struct reloc_t *reloc = RZ_NEW0(struct reloc_t);
+		if (!reloc) {
+			free(info);
+			free(sym_name);
+			return;
+		}
+
+		reloc->addr = MACH0_(paddr_to_vaddr)(bin, a_info.r_address);
+		reloc->offset = a_info.r_address;
+		reloc->ord = sym_num;
+		reloc->type = a_info.r_type; // enum RelocationInfoType
+		reloc->external = a_info.r_extern;
+		reloc->pc_relative = a_info.r_pcrel;
+		reloc->size = 1 << a_info.r_length; // macho/reloc.h says: 0=byte, 1=word, 2=long, 3=quad
+		rz_str_ncpy(reloc->name, sym_name, sizeof(reloc->name) - 1);
+		rz_skiplist_insert(relocs, reloc);
+		free(sym_name);
+	}
+	free(info);
+}
+
+static bool is_valid_ordinal_table_size(ut64 size) {
+	return size > 0 && size <= UT16_MAX;
+}
+
+static int parse_import_ptr(struct MACH0_(obj_t) * bin, struct reloc_t *reloc, int idx) {
+	int i, j, sym;
+	size_t wordsize;
+	ut32 stype;
+	wordsize = get_word_size(bin);
+	if (idx < 0 || idx >= bin->nsymtab) {
+		return 0;
+	}
+	if ((bin->symtab[idx].n_desc & REFERENCE_TYPE) == REFERENCE_FLAG_UNDEFINED_LAZY) {
+		stype = S_LAZY_SYMBOL_POINTERS;
+	} else {
+		stype = S_NON_LAZY_SYMBOL_POINTERS;
+	}
+
+	reloc->offset = 0;
+	reloc->addr = 0;
+	reloc->addend = 0;
+#define CASE(T) \
+	case ((T) / 8): reloc->type = RZ_BIN_RELOC_##T; break
+	switch (wordsize) {
+		CASE(8);
+		CASE(16);
+		CASE(32);
+		CASE(64);
+	default: return false;
+	}
+#undef CASE
+
+	for (i = 0; i < bin->nsects; i++) {
+		if ((bin->sects[i].flags & SECTION_TYPE) == stype) {
+			for (j = 0, sym = -1; bin->sects[i].reserved1 + j < bin->nindirectsyms; j++) {
+				int indidx = bin->sects[i].reserved1 + j;
+				if (indidx < 0 || indidx >= bin->nindirectsyms) {
+					break;
+				}
+				if (idx == bin->indirectsyms[indidx]) {
+					sym = j;
+					break;
+				}
+			}
+			reloc->offset = sym == -1 ? 0 : bin->sects[i].offset + sym * wordsize;
+			reloc->addr = sym == -1 ? 0 : bin->sects[i].addr + sym * wordsize;
+			return true;
+		}
+	}
+	return false;
+}
+
+RZ_BORROW RzSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) * bin) {
```

Comment:
This function is HUGE... What about splitting it up while you are on it?

Reply:
I think it is too fragile. I would rather do that in another pr.